Fix start cluster when docker container is connected to several networks

fredericgermain opened the pull request:
fixes #10079

I followed this tutorial and connected minikube to an extra network for metallb : https://www.hashicorp.com/blog/deploy-consul-cluster-peering-locally-with-minikube

- Before: request made in pkg/drivers/kic/oci/network.go by dockerContainerIP : 

`docker container inspect -f "{{range .NetworkSettings.Networks}}{{.IPAddress}},{{.GlobalIPv6Address}}{{end}}" dc1`,   

would display something like `192.168.121.2,172.68.0.2,`.

- After: I just changed the request to just ask the ip of the main network
 
`docker container inspect -f '{{with (index .NetworkSettings.Networks "dc1")}}{{.IPAddress}},{{.GlobalIPv6Address}}{{end}}' dc1`

Now only the addresses of the main network are returned `192.168.121.2,`.

medyagh commented:
@fredericgermain thank you very much for this contribution and sharing the tutorial, 
do you mind sharing the Steps to to do to recreate the environment that the bug happens ?

I think this could be added as an integration test or maybe unit test, would you like to add an integration test for this ?

medyagh commented:
/ok-to-test

medyagh commented:
[fredericgermain](https://github.com/fredericgermain) the Docker Linux test are failing and related to this PR https://storage.googleapis.com/minikube-builds/logs/18943/34582/Docker_Linux.html

fredericgermain commented:
That's a lot of errors indeed!
I will try to have a look within the next two weeks.

fredericgermain commented:
Hi @medyagh 

I added a integration test extnet_test.go and fixed the errors linked by my previous commit.
This test only apply on driver "docker"

In my "not so proud" things in my commit, that worths maybe some talk : 
- New DriverArg() : I wanted to test if --driver argument is given, or is default driver. DockerDriver() do not say if no --driver was given and I was in that case when running `TEST_FILES=extnet_test.go make integration` first.
- Not sure how to clean the external network properly in the test. If the network is still here a second run of test will fails as I give fixed ip range to the network

I also have some other small commit that fix some tests. But some tests just fails randomly, I guess it's a known problem?

medyagh commented:
@prezha @ComradeProgrammer do you mind reviewing this PR

prezha commented on test/integration/extnet_test.go:
```diff
@@ -0,0 +1,258 @@
+}
+
+// deleteExtnet just starts a new minikube cluster
+func deleteExtnet(ctx context.Context, t *testing.T, profile string, extnetNetworkName string) {
+	cmd := exec.CommandContext(ctx, "docker", "network", "rm", extnetNetworkName)
+
+	result, err := Run(t, cmd)
+	if err != nil {
+		t.Fatalf("failed to execute 'docker network delete', error: %v, output: %s", err, result.Output())
+	}
+	extnetNetworkID := result.Output()
+	fmt.Fprintf(os.Stderr, "%s", extnetNetworkID)
+}
```
as commented above, this becomes our deferred function called when tests ends (regardless how):

```
func deleteExtnet(t *testing.T, extNetworkName string) {
	ctx, cancel := context.WithTimeout(context.Background(), Seconds(10))
	defer cancel()

	cmd := exec.CommandContext(ctx, "docker", "network", "rm", extNetworkName)

	result, err := Run(t, cmd)
	if err != nil {
		t.Fatalf("failed to execute 'docker network delete', error: %v, output: %s", err, result.Output())
	}
	t.Logf("deleted new network %q", extNetworkName)
}
```

fredericgermain replied:
I did the defer trick, see below

prezha reviewed (changes requested):
hey @fredericgermain, thanks so much for taking the time to make minikube better!

i left some comments i think could help improve your pr and then merge it - please have a look and feel free to ask any clarifying questions

also, if you want to to test your changes locally, you can use:
- `make clean && make && make test`
- `make integration -e TEST_ARGS="-minikube-start-args='--driver=docker --container-runtime=docker --alsologtostderr -v=7' -test.run TestExtNet"`
- `make integration -e TEST_ARGS="-minikube-start-args='--driver=docker --container-runtime=docker --alsologtostderr -v=7' -test.run TestMultiNode"`

prezha commented on pkg/drivers/kic/oci/network.go:
```diff
@@ -225,9 +225,10 @@ func podmanContainerIP(ociBin string, name string) (string, string, error) {
+func dockerContainerIP(ociBin string, containerNameOrID string, networkID string) (string, string, error) {
 	// retrieve the IP address of the node using docker inspect
-	lines, err := inspect(ociBin, name, "{{range .NetworkSettings.Networks}}{{.IPAddress}},{{.GlobalIPv6Address}}{{end}}")
+	format := fmt.Sprintf("{{with (index .NetworkSettings.Networks \"%s\")}}{{.IPAddress}},{{.GlobalIPv6Address}}{{end}}", networkID)
```
we can use `%q` instead of `\"%s\"`

fredericgermain replied:
changed, fixed

fredericgermain commented:
Hi everyone!

Thank you for the comprehensive review.
I'll do my best to have the points raised fixed by the end of the week.

fredericgermain commented on pkg/drivers/kic/kic.go:
```diff
@@ -293,7 +293,12 @@ func (d *Driver) DriverName() string {
+        networkName := d.NodeConfig.Network
+        if networkName == "" {
+                networkName = d.NodeConfig.ClusterName
+        }
```
fixed

fredericgermain commented on pkg/drivers/kic/oci/network.go:
```diff
@@ -197,25 +197,25 @@ func ForwardedPort(ociBin string, ociID string, contPort int) (int, error) {
 
 // ContainerIPs returns ipv4,ipv6, error of a container by their name
-func ContainerIPs(ociBin string, name string) (string, string, error) {
+func ContainerIPs(ociBin string, containerNameOrID string, networkID string) (string, string, error) {
```
fixed

fredericgermain commented on test/integration/extnet_test.go:
```diff
@@ -0,0 +1,258 @@
+func fail(ctx context.Context, t *testing.T, profile string, extnetNetworkName string) {
+	defer PostMortemLogs(t, profile)
+  t.Errorf("testing failure")
+}
```
removed => fixed

fredericgermain commented on test/integration/extnet_test.go:
```diff
@@ -0,0 +1,258 @@
+	})
+}
+
+// connectExtnet create a docker network
```
typo fixed

fredericgermain commented on test/integration/extnet_test.go:
```diff
@@ -0,0 +1,258 @@
+	fmt.Fprintf(os.Stderr, "%s", extnetNetworkID)
+}
+
+// connectExtnet connect network to the minikube cluster
```
typo fixed

fredericgermain commented on test/integration/extnet_test.go:
```diff
@@ -0,0 +1,258 @@
+	args := []string{"start", "-p", profile, "--alsologtostderr", "-v=5"}
+	rr, err := Run(t, exec.CommandContext(ctx, Target(), args...))
+	if err != nil {
+		t.Errorf("failed to start minikube with args: %q : %v", rr.Command(), err)
```
fixed

fredericgermain commented on test/integration/extnet_test.go:
```diff
@@ -0,0 +1,258 @@
+	fmt.Fprintf(os.Stderr, "%s", extnetNetworkID)
+}
+
+// deleteExtnet just starts a new minikube cluster
```
fixed

fredericgermain commented on test/integration/extnet_test.go:
```diff
@@ -0,0 +1,258 @@
+	}
+}
+
+// extnetValidateStop  runs minikube Stop 
```
fixed

fredericgermain commented on test/integration/helpers_test.go:
```diff
@@ -687,3 +688,20 @@ func CopyDir(src, dst string) error {
+	}
+	return rr.Stdout.Bytes()
+}
+
```
see comment belove about how I run the test, and how I think this function used in several test should be moved to the helper files

fredericgermain commented on test/integration/main_test.go:
```diff
@@ -149,6 +149,18 @@ func PodmanDriver() bool {
+	}
+	return "", false
+}
+
```
reverted that part indeed. when I started this, my default driver is docker, and I didn't know about the -minikube-start-args argument to pass driver to tests.

fredericgermain commented on test/integration/extnet_test.go:
```diff
@@ -0,0 +1,258 @@
+	t.Logf("running with runtime:%s DriverArg:%s goos:%s goarch:%s", ContainerRuntime(), driverArg, runtime.GOOS, runtime.GOARCH)
+	if driverArgPresent && driverArg != "docker" {
+		t.Skip("skipping: only docker driver supported")
+	}
```
reverted as said just before

fredericgermain commented on test/integration/extnet_test.go:
```diff
@@ -0,0 +1,258 @@
+	case "crio":
+	    t.Skip("skipping: access direct test is broken on windows: https://github.com/kubernetes/minikube/issues/8304")
+	case "docker":
+	}
```
Indeed, I tested with all these container runtime configs and it works. I removed these lines

fredericgermain commented on test/integration/extnet_test.go:
```diff
@@ -0,0 +1,258 @@
+
+	extnetNetworkName := fmt.Sprintf("%s-%s", "network-extnet", fmt.Sprintf("%06d", time.Now().UnixNano()%1000000))
+
+	// Serial tests
```
removed => fixed

fredericgermain commented on test/integration/extnet_test.go:
```diff
@@ -0,0 +1,258 @@
+func createExtnet(ctx context.Context, t *testing.T, profile string, extnetNetworkName string) {
+	subnet := "172.28.0.0/16"
+	ipRange :="172.28.0.0/24"
+	cmd := exec.CommandContext(ctx, "docker", "network", "create", extnetNetworkName, fmt.Sprintf("--subnet=%s", subnet), fmt.Sprintf("--ip-range=%s", ipRange))
```
removed the hard-coding => fixed

fredericgermain commented on test/integration/extnet_test.go:
```diff
@@ -0,0 +1,258 @@
+		t.Fatalf("failed to execute 'docker network create', error: %v, output: %s", err, result.Output())
+	}
+	extnetNetworkID := result.Output()
+	fmt.Fprintf(os.Stderr, "%s", extnetNetworkID)
```
removed or changed all the fmt.Fprintf with t.Logf

fredericgermain commented on test/integration/extnet_test.go:
```diff
@@ -0,0 +1,258 @@
+)
+
+// TestExtNet tests minikube external network functionality
+func TestExtNet(t *testing.T) {
```
I did change to multinetwork
I didn't prefix with kic. Even though the support for podman could be added relatively easily, if these external networks find more use, maybe this test could even be ported on qemu, kvm, ...

fredericgermain commented on test/integration/extnet_test.go:
```diff
@@ -0,0 +1,258 @@
+				if t.Failed() && *postMortemLogs {
+					PostMortemLogs(t, profile)
+				}
+			})
```
Although I agree, see my note about how I mostly copied pause_test.go, and I didn't dare to change this out of consistency with pause_test.go

fredericgermain commented on test/integration/extnet_test.go:
```diff
@@ -0,0 +1,258 @@
+			{"Start", extnetValidateStart},
+			{"Delete", extnetValidateDelete},
+		//	{"Fail", fail},
+			{"VerifyDeletedResources", extnetValidateVerifyDeleted},
```
I copied this cleaning principle from pause_test.go. I think it's quite nice to test the deletion with a very controlled way. And the Cleanup is called effectively in case of early error. I made some change to be closer to this principle (will talk about this in the main message)

fredericgermain commented on test/integration/extnet_test.go:
```diff
@@ -0,0 +1,258 @@
+			{"FreshStart", extnetValidateFreshStart},
+			{"ConnectExtnet", connectExtnet},
+			{"Stop", extnetValidateStop},
+			{"VerifyStatus", extnetValidateStatus},
```
cf message before and copy of pause_test.go principles

fredericgermain commented on test/integration/extnet_test.go:
```diff
@@ -0,0 +1,258 @@
+		}
+	}
+
+}
```
cf message before and copy of pause_test.go principles

fredericgermain commented on test/integration/extnet_test.go:
```diff
@@ -0,0 +1,258 @@
+	if cs.StatusName != "Stopped" {
+		t.Fatalf("incorrect status name: %v", cs.StatusName)
+	}
+}
```
cf message before and copy of pause_test.go principles

fredericgermain commented on test/integration/extnet_test.go:
```diff
@@ -0,0 +1,258 @@
+	if err != nil {
+		t.Errorf("failed to delete minikube with args: %q : %v", rr.Command(), err)
+	}
+}
```
cf message before and copy of pause_test.go principles

fredericgermain commented on test/integration/extnet_test.go:
```diff
@@ -0,0 +1,258 @@
+	if err != nil {
+		t.Errorf("failed to start minikube with args: %q : %v", rr.Command(), err)
+	}
+}
```
done => fixed?

fredericgermain commented on test/integration/extnet_test.go:
```diff
@@ -0,0 +1,258 @@
+			if t.Failed() && *postMortemLogs {
+				PostMortemLogs(t, profile)
+			}
+		})
```
cf message before and copy of pause_test.go principles.
I kinda duplicated the delete of external network. deleteExtnet is the "controlled" deletion, and CleanupExtnet is the deletion to be done in defer

fredericgermain commented on test/integration/extnet_test.go:
```diff
@@ -0,0 +1,258 @@
+	args := []string{"stop", "-p", profile, "--alsologtostderr", "-v=5"}
+	rr, err := Run(t, exec.CommandContext(ctx, Target(), args...))
+	if err != nil {
+		t.Errorf("failed to stop minikube with args: %q : %v", rr.Command(), err)
```
fixed

fredericgermain commented on test/integration/extnet_test.go:
```diff
@@ -0,0 +1,258 @@
+	if err != nil {
+		t.Fatalf("failed to execute 'docker network create', error: %v, output: %s", err, result.Output())
+	}
+	extnetNetworkID := result.Output()
```
just using extnetNetworkName in logs

fredericgermain commented on test/integration/extnet_test.go:
```diff
@@ -0,0 +1,258 @@
+	if err != nil {
+		t.Fatalf("failed to execute 'docker network connect', error: %v, output: %s", err, result.Output())
+	}
+	extnetNetworkID := result.Output()
```
indeed, changed

fredericgermain commented:
Hi @prezha !

Thank you so much for you very detailed review, it's really appreciated. I tried to work my best so hopefully you do not have to go back at this too many times.

Unfortunately I renamed extnet_test.go to multinetwork_test.go, which means the fixes are not going to be easily seen in github UI.

I do agree with most of comments and made the changes accordingly.

My main concern is about removing all the Delete* functions that are part of the test. I mostly copied that file from pause_test.go, and there deletes and checks are made as part of the test. I think it's kind of nice to test the removal in a control way. But it's also maybe unnecessary. Let me know if you want to stick with your idea of removing all these delete/check functions in this test, or if we stick to the principles of pause_test.go

I did  change and clean the way the tests are run, and made a CleanuExtnet in the defer as suggested : it's indeed much better than my tricks.

---
About the moving of runStatusCmd :
Although I guess it's not used anymore, I use to run my test with `time env -- TEST_ARGS="-minikube-start-args \"--driver=docker \"" TEST_FILES="multinetwork_test.go" make integration`
Using TEST_FILES as a filter, we cannot use functions from other tests, and we can only use function for the helpers/utilitu files.
That's why I moved runStatusCmd, which was already used by both pause_test.go and status_test.go, into helpers_test.go

prezha commented on test/integration/multinetwork_test.go:
```diff
@@ -0,0 +1,295 @@
+
+// CleanupExtnet removes the external network in docker, no error on failure, used for Cleanup
+func CleanupExtnet(t *testing.T) {
+	ctx2, cancel2 := context.WithTimeout(context.Background(), Seconds(10))
```
nit: ctx and cancel are local variables to this function, so no need to have suffix, but that's ok

fredericgermain replied:
fixed

prezha reviewed (commented):
hey @fredericgermain thanks for addressing those comments!

i've left few more, mostly around handling errors instead of ignoring them - could be useful in case we need to debug tests that fail

as for \*Delete\* tests, i think they are not _necessary_ they are also not_wrong_, but we could "enrich" them by adding what we care about here specifically, and that's what happens with the external/added network

finally, regarding moving of runStatusCmd: i get why you did that to run this specific test, but i'm not sure if that can perhaps break some other test, maybe not

thanks for persisting, pls have a look at the new comments and then we'll re-run all the tests and see how we stand, but i think we're getting there!

fredericgermain commented on test/integration/multinetwork_test.go:
```diff
@@ -0,0 +1,295 @@
+	})
+}
+
+// connectExtnet creates a docker network
```
fixed

fredericgermain commented on test/integration/multinetwork_test.go:
```diff
@@ -0,0 +1,295 @@
+		if PodmanDriver() {
+			bin = "podman"
+		}
+		ipv4, ipv6, _ := oci.ContainerIPs(bin, profile, profile)
```
fixed

fredericgermain commented on test/integration/multinetwork_test.go:
```diff
@@ -0,0 +1,295 @@
+		if ipv6 != clusterIPv6 {
+			t.Fatalf("clusterIPv6 mismatch %s != %s", clusterIPv6, ipv6)
+		}
+		ipv4, ipv6, _ = oci.ContainerIPs(bin, profile, extnetNetworkName)
```
fixed

fredericgermain commented:
Hi @prezha !

Again; thank you for your time to review this.

For PostMortemLogs, I aligned with the way it is do in image_test.go

I moved VerifyDeletedResources at the very end and it checks the external network is removed.

I didn't remove extnetValidateStop. Although I agree it shouldn't be really done in this test, a test for the stop status is not done in any other test.
But as suggested, I added a test to check that the network config is what expected after a restart.

fredericgermain commented:
Hi @prezha Hi @medyagh !

I guess there was a busy period. I was wondering if you had time to have a look at the lastest version?
I can do a rebase as it starts to be quite old.
